### PR TITLE
configure reset partition boot

### PR DESCRIPTION
**Pull request by mwhudson**
It would be good to figure out how to test this automatically (I've tested it by hand and it seems to work)

### CODE REVIEW

File: subiquity/server/controllers/install.py [hunk 1]
```diff
@@ -380,6 +385,61 @@ async def run_curtin_step(name, stages, step_config, source=None):
...
+
+    @with_context(description="creating boot entry for reset partition")
+    async def create_rp_boot_entry(self, context, rp):
+        fs_controller = self.app.controllers.Filesystem
+        if not fs_controller.reset_partition_only:
+            cp = await self.app.command_runner.run(
+                ['lsblk', '-n', '-o', 'UUID', rp.path],
+                capture=True)
+            uuid = cp.stdout.decode('ascii').strip()
+            conf = grub_reset_conf.replace("#UUID#", uuid)
```
Comment (dbungert) [inline subiquity/server/controllers/install.py:398]: nitpick: suggest using .format instead, can do this in a single pass

File: subiquity/server/controllers/install.py [hunk 2]
```diff
@@ -380,6 +385,61 @@ async def run_curtin_step(name, stages, step_config, source=None):
...
+            conf = conf.replace("#PARTITION#", str(rp.number))
+            with open(self.tpath('etc/grub.d/99_reset'), 'w') as fp:
```
Comment (dbungert) [inline subiquity/server/controllers/install.py:400]: I like the convention that files we generate on the target system get a header that helps the end user figure out how they got there.  Something akin to `generate_timestamped_header()` would be nice.

File: subiquity/server/controllers/install.py [hunk 3]
```diff
@@ -380,6 +385,61 @@ async def run_curtin_step(name, stages, step_config, source=None):
...
+            return
+        new_bootnum = new_bootnums.pop()
+        new_entry = efi_state_after.entries[new_bootnum]
+        was_dup = False
+        for entry in efi_state_before.entries.values():
+            if entry.path == new_entry.path and entry.label == new_entry.label:
+                was_dup = True
+        if was_dup:
+            cmd = [
+                'efibootmgr', '--delete-bootnum',
```
Comment (dbungert) [inline subiquity/server/controllers/install.py:434]: I'd like some unit tests on this dup-cleanup functionality.  I want to reduce the likelihood of the wrong boot entries getting deleted.

File: subiquity/server/controllers/install.py [hunk 4]
```diff
@@ -590,3 +653,15 @@ async def stop_unattended_upgrades(self):
         "${distro_id}ESM:${distro_codename}-infra-security";
 };
 """
+
+grub_reset_conf = """\
+#!/bin/bash -e
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/controllers/install.py:658]: yes probably.

File: subiquity/server/controllers/install.py [hunk 5]
```diff
@@ -380,6 +385,61 @@ async def run_curtin_step(name, stages, step_config, source=None):
...
+            '--loader', '\\EFI\\boot\\shimx64.efi',
+            '--disk', rp.device.path,
+            '--part', str(rp.number),
+            '--label', "Restore Ubuntu to factory state",
+            ]
+        await self.app.command_runner.run(cmd)
+        efi_state_after = get_efibootmgr('/')
+        new_bootnums = (
+            set(efi_state_after.entries) - set(efi_state_before.entries))
+        if len(new_bootnums) == 0:
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/controllers/install.py:424]: hmmf I have my own biases against general boolean testing of things that are not booleans but well. OK :)

File: subiquity/server/controllers/install.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/install.py:398]: yeah that would make sense

**Review by dbungert — APPROVED** [review]

**Pull request merged**